`SchemaFactory` bound to Jackson

The changes implementing enhancement #245 allow libraries layering on SmallRye OpenAPI to implement `AnnotationScannerExtension` in order to, among other things, provide their own parsing rather than Jackson.

One of our users seems to have found a spot where SmallRye is still tightly bound to Jackson. This happens to be in 2.1.16.

It looks as if adding `@Schema` to a field triggers this, in that the lambda in `SchemaFactory#parseSchemaAttr` invokes `JsonUtil#parseValue` which invokes Jackson.

Below is the relevant part of a stack trace.

Not sure if there might be other similar places in the code; I didn't see others in `SchemaFactory` but I didn't look beyond there.

```
Caused by: java.lang.NoClassDefFoundError: com/fasterxml/jackson/databind/JsonNode
	at io.smallrye.openapi.runtime.io.schema.SchemaFactory.lambda$parseSchemaAttr$5(SchemaFactory.java:315)
	at io.smallrye.openapi.runtime.io.schema.SchemaFactory.readAttr(SchemaFactory.java:343)
	at io.smallrye.openapi.runtime.io.schema.SchemaFactory.parseSchemaAttr(SchemaFactory.java:309)
	at io.smallrye.openapi.runtime.io.schema.SchemaFactory.readSchema(SchemaFactory.java:180)
	at io.smallrye.openapi.runtime.scanner.dataobject.AnnotationTargetProcessor.readSchemaAnnotatedField(AnnotationTargetProcessor.java:327)
	at io.smallrye.openapi.runtime.scanner.dataobject.AnnotationTargetProcessor.processField(AnnotationTargetProcessor.java:163)
	at io.smallrye.openapi.runtime.scanner.dataobject.AnnotationTargetProcessor.process(AnnotationTargetProcessor.java:65)
	at io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.lambda$depthFirstGraphSearch$1(OpenApiDataObjectScanner.java:262)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1845)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.depthFirstGraphSearch(OpenApiDataObjectScanner.java:262)
	at io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.process(OpenApiDataObjectScanner.java:213)
	at io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.process(OpenApiDataObjectScanner.java:159)
	at io.smallrye.openapi.runtime.io.schema.SchemaFactory.introspectClassToSchema(SchemaFactory.java:527)
	at io.smallrye.openapi.runtime.io.schema.SchemaFactory.typeToSchema(SchemaFactory.java:444)
	at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.lambda$processClassSchemas$3(OpenApiAnnotationScanner.java:208)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.processClassSchemas(OpenApiAnnotationScanner.java:208)
	at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.scanMicroProfileOpenApiAnnotations(OpenApiAnnotationScanner.java:153)
	at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.scan(OpenApiAnnotationScanner.java:118)
	at io.helidon.openapi.OpenAPISupport.lambda$expandModelUsingAnnotations$3(OpenAPISupport.java:364)
```

--- a/core/src/main/java/io/smallrye/openapi/runtime/io/schema/SchemaFactory.java
+++ b/core/src/main/java/io/smallrye/openapi/runtime/io/schema/SchemaFactory.java
@@ -34,7 +34,6 @@ import io.smallrye.openapi.api.models.media.SchemaImpl;
 import io.smallrye.openapi.api.util.MergeUtil;
 import io.smallrye.openapi.runtime.io.CurrentScannerInfo;
 import io.smallrye.openapi.runtime.io.IoLogging;
-import io.smallrye.openapi.runtime.io.JsonUtil;
 import io.smallrye.openapi.runtime.io.extension.ExtensionReader;
 import io.smallrye.openapi.runtime.io.externaldocs.ExternalDocsConstant;
 import io.smallrye.openapi.runtime.io.externaldocs.ExternalDocsReader;
@@ -181,7 +180,7 @@ public class SchemaFactory {
         schema.setExternalDocs(ExternalDocsReader.readExternalDocs(context, externalDocsAnnotation));
         schema.setDeprecated(readAttr(annotation, SchemaConstant.PROP_DEPRECATED, defaults));
         schema.setType(readSchemaType(annotation, schema, defaults));
-        schema.setExample(parseSchemaAttr(annotation, SchemaConstant.PROP_EXAMPLE, defaults, schema.getType()));
+        schema.setExample(parseSchemaAttr(context, annotation, SchemaConstant.PROP_EXAMPLE, defaults, schema.getType()));
         schema.setDefaultValue(readAttr(annotation, SchemaConstant.PROP_DEFAULT_VALUE, defaults));
         schema.setDiscriminator(
                 readDiscriminator(context,
@@ -314,21 +313,28 @@ public class SchemaFactory {
      * they key. Array-typed annotation values will be converted to List.
      *
      * @param <T> the type of the annotation attribute value
+     * @param context scanning context
      * @param annotation the annotation to read
      * @param propertyName the name of the attribute to read
      * @param defaults map of default values
      * @param schemaType related schema type for this attribute
      * @return the annotation attribute value, a default value, or null
      */
-    static Object parseSchemaAttr(AnnotationInstance annotation, String propertyName, Map<String, Object> defaults,
-            SchemaType schemaType) {
+    static Object parseSchemaAttr(AnnotationScannerContext context, AnnotationInstance annotation, String propertyName,
+            Map<String, Object> defaults, SchemaType schemaType) {
         return readAttr(annotation, propertyName, value -> {
             if (!(value instanceof String)) {
                 return value;
             }
             String stringValue = ((String) value);
             if (schemaType != SchemaType.STRING) {
-                return JsonUtil.parseValue(stringValue);
+                Object parsedValue;
+                for (AnnotationScannerExtension e : context.getExtensions()) {
+                    parsedValue = e.parseValue(stringValue);
+                    if (parsedValue != null) {
+                        return parsedValue;
+                    }
+                }
             }
             return stringValue;
         }, defaults);

--- a/pom.xml
+++ b/pom.xml
@@ -21,7 +21,7 @@
         <jackson-bom.version>2.13.3</jackson-bom.version>
         <version.eclipse.microprofile.config>3.0</version.eclipse.microprofile.config>
         <version.io.smallrye.smallrye-config>3.0.0-RC2</version.io.smallrye.smallrye-config>
-        <version.eclipse.microprofile.openapi>3.1-SNAPSHOT</version.eclipse.microprofile.openapi>
+        <version.eclipse.microprofile.openapi>3.1-RC1</version.eclipse.microprofile.openapi>
         <version.org.hamcrest>1.3</version.org.hamcrest>
         <version.org.hamcrest.java-hamcrest>2.0.0.0</version.org.hamcrest.java-hamcrest>
         <version.org.skyscreamer>1.5.0</version.org.skyscreamer>

